### DISCUSSION TITLE
JWT auth with custom backend access token never update till the client make hard refresh

### DISCUSSION THREAD

**Discussion by Elabbasy00** [nextauthjs/next-auth#6008] [Help]
### Environment

  System:
    OS: Linux 5.15 Ubuntu 22.04.1 LTS 22.04.1 LTS (Jammy Jellyfish)
    CPU: (8) x64 Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
    Memory: 9.52 GB / 15.58 GB
    Container: Yes
    Shell: 5.1.16 - /bin/bash
  Binaries:
    Node: 16.15.0 - ~/.nvm/versions/node/v16.15.0/bin/node
    Yarn: 1.22.18 - /usr/bin/yarn
    npm: 8.5.5 - ~/.nvm/versions/node/v16.15.0/bin/npm
  Browsers:
    Chrome: 108.0.5359.94
    Firefox: 107.0.1
  npmPackages:
    next: 13.0.6 => 13.0.6 
    next-auth: ^4.18.3 => 4.18.3 
    react: 18.2.0 => 18.2.0 

### Reproduction URL

null

### Describe the issue

**I use CredentialsProvider to auth users with the Django backend
take the Credentials in authorizeCallback**

```javascript
 async authorize(credentials, req) {
        try {
          const respo = await axios.post(
            "http://127.0.0.1:8000/auth/jwt/create/",
            {
             // hard coded for testing
              phone_number: "123456789",
              password: "123456",
            }
          );

          const { access, refresh } = respo.data;

          const respo1 = await axios.get(
            "http://127.0.0.1:8000/auth/users/me/",
            {
              headers: { Authorization: "JWT " + access },
            }
          );

          return { ...respo1.data, refresh: refresh, access: access };
        } catch (e) {
          return false;
        }
},
```

**request django to create access, refresh token and user info 
then check for access token expire in JWTCallback**

``` javascript
    async jwt({ token, user, account, profile, isNewUser }) {
      if (user) {
        token.accessToken = user.access;
        token.refreshToken = user.refresh;

        return token;
      }

      if (isJwtExpired(token.accessToken)) {
        const [newAccessToken, newRefreshToken] = await refreshToken(
          token.refreshToken
        );

        if (newAccessToken) {
          token = {
            ...token,
            accessToken: newAccessToken,
            iat: Math.floor(Date.now() / 1000),
            exp: Math.floor(Date.now() / 1000 + 2 * 60 * 60),
          };

          return token;
        }
        setAxiosAuthToken(null);
        return {
          ...token,
          exp: 0,
          error: "RefreshAccessTokenError",
        };
      }

      // token valid
      return token;
    },
```
**session callback call setAxiosAuthToken to add Jwt in every Axios request**

```javascript
async session({ session, user, token }) {
      session.accessToken = token.accessToken;
      session.error = token.error;
      setAxiosAuthToken(session.accessToken);
      return session;
    }
},
```
the problem is when the access token is refreshed setAxiosAuthToken never gets an update
and still request the old access token

I tried using `getSession()` in Axios `interceptors.response` but `getSession()` always returns null even in `interceptors.request`
also tried requesting session URL in Axios interceptors but respond with {}

Then I found an article for someone explaining the problem and offering a solution which is to use useSwr to keep the session up to date by pinging the session URL in interval

```javascript
  const { data, error } = useSwr(sessionUrl, fetchSession, {
    revalidateOnFocus: true,
    revalidateOnMount: true,
    revalidateOnReconnect: true,
  });

  useEffect(() => {
    const intervalId = setInterval(
      () => mutate(sessionUrl),
      (refreshInterval || 20) * 1000
    );

    return () => clearInterval(intervalId);
  }, []);

  return {
    session: data,
    loading: typeof data === "undefined" && typeof error === "undefined",
  };
}
```

HOc for require auth views

```javascript
export function withAuth(refreshInterval) {
  return function (Component) {
    return function (props) {
      const { session, loading } = useAuth(refreshInterval);

      if (typeof window !== undefined && loading) {
        return null;
      }

      if (
        (!loading && !session) ||
        session?.error === "RefreshAccessTokenError"
      ) {
        return (
          <>
            <button onClick={() => signIn()}>Sign in</button>
          </>
        );
      }
      return <Component session={session} {...props} />;
    };
  };
}


```

**but the problem is remain, session updated with the latest access Token but Axios never gets updated till the next hard refresh**

t

### How to reproduce

trying to use JWT CredentialsProvider  with Custom backend, next-auth@4.18.3  
session only update by hard refreshing

### Expected behavior

update the session, axios auth token with latest JWT accessToken after refreshing it from backend

**Comment by uibprodo** [comment]
Did you solve this issue? facing the same thing
